plugins/usbdmx/JaRuleWidget.cpp

```diff
@@ -18,50 +18,253 @@
+  : m_executor(executor),
+    m_adaptor(adaptor),
+    m_device(usb_device),
+    m_usb_handle(NULL),
+    m_uid(0, 0) {
+  m_adaptor->RefDevice(m_device);
+}
+
+JaRuleWidget::~JaRuleWidget() {
+  STLDeleteElements(&m_ports);
+
+  if (m_usb_handle) {
+    m_adaptor->Close(m_usb_handle);
+  }
+
+  m_adaptor->UnrefDevice(m_device);
 }
 
 bool JaRuleWidget::Init() {
-  return m_widget_impl->Init();
+  bool ok = InternalInit();
+  if (!ok) {
+    STLDeleteElements(&m_ports);
+    if (m_usb_handle) {
+      m_adaptor->Close(m_usb_handle);
+      m_usb_handle = NULL;
+    }
+  }
+  return ok;
+}
+
+void JaRuleWidget::CancelAll(uint8_t port_index) {
+  if (port_index > m_ports.size() - 1) {
+    return;
+  }
+  JaRuleWidgetPort *port_info = m_ports[port_index];
+  port_info->CancelAll();
+}
+
+uint8_t JaRuleWidget::PortCount() const {
+  return m_ports.size();
+}
+
+UID JaRuleWidget::GetUID() const {
+  return m_uid;
+}
+
+string JaRuleWidget::ManufacturerString() const {
+  return m_manufacturer;
 }
 
-void JaRuleWidget::SendRDMRequest(RDMRequest *request,
-                                  ola::rdm::RDMCallback *on_complete) {
-  m_queueing_controller.SendRDMRequest(request, on_complete);
+string JaRuleWidget::ProductString() const {
+  return m_product;
 }
 
-void JaRuleWidget::RunFullDiscovery(RDMDiscoveryCallback *callback) {
-  m_queueing_controller.RunFullDiscovery(callback);
+JaRulePortHandle* JaRuleWidget::ClaimPort(uint8_t port_index) {
+  if (port_index > m_ports.size() - 1) {
+    return NULL;
+  }
+  return m_ports[port_index]->ClaimPort();
 }
 
-void JaRuleWidget::RunIncrementalDiscovery(RDMDiscoveryCallback *callback) {
-  m_queueing_controller.RunIncrementalDiscovery(callback);
+void JaRuleWidget::ReleasePort(uint8_t port_index) {
+  if (port_index > m_ports.size() - 1) {
+    return;
+  }
+  m_ports[port_index]->ReleasePort();
 }
 
-bool JaRuleWidget::SendDMX(const DmxBuffer &buffer) {
-  return m_widget_impl->SendDMX(buffer);
+void JaRuleWidget::SendCommand(uint8_t port_index,
+                               CommandClass command,
+                               const uint8_t *data,
+                               unsigned int size,
+                               CommandCompleteCallback *callback) {
+  if (port_index > m_ports.size() - 1) {
+    OLA_WARN << "Invalid JaRule Port " << static_cast<int>(port_index);
+    if (callback) {
+      callback->Run(jarule::COMMAND_RESULT_INVALID_PORT, 0, 0, ByteString());
+    }
+    return;
+  }
+  m_ports[port_index]->SendCommand(command, data, size, callback);
 }
 
-void JaRuleWidget::ResetDevice() {
-  m_widget_impl->ResetDevice();
+// Private Methods
+// ----------------------------------------------------------------------------
+
+bool JaRuleWidget::InternalInit() {
+  struct libusb_config_descriptor *config;
+  int error = m_adaptor->GetActiveConfigDescriptor(m_device, &config);
+  if (error) {
+    OLA_WARN << "Failed to get active config descriptor for Ja Rule device: "
```

Comment:
Can you add a reference as to which device, in case we have more than one connected.